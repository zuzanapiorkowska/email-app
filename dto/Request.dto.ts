import "reflect-metadata";
import { Type } from "class-transformer";
import {
  IsString,
  IsInt,
  Min,
  Max,
  IsEmail,
  ValidateNested,
  IsArray,
} from "class-validator";

export class Answer {
  @IsString()
  answer!: string;

  @IsInt()
  @Min(1)
  @Max(5)
  choice!: number;
}

export class Request {
  @IsString()
  @IsEmail()
  email!: string;

  @ValidateNested({ each: true })
  @Type(() => Answer)
  @IsArray()
  answers!: Answer[];
}
