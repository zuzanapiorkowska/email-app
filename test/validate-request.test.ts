import { validate } from "class-validator";
import { Answer, Request } from "../dto/Request.dto";

function generateRequest(overrides: { email?: any; answers?: any }) {
  const request = new Request();
  request.email =
    overrides.email === undefined ? "mihau@gmail.com" : overrides.email;
  request.answers = overrides.answers === undefined ? [] : overrides.answers;

  return request;
}

// test('passes with correct request', async () => {
//   const request = generateRequest({});
//   expect(await validate(request)).toEqual([])

test("passes with correct request", async () => {
  const request = generateRequest({});
  expect(await validate(request)).toEqual([]);
});

test("fails without email", async () => {
  const request = generateRequest({ email: null });
  expect(await validate(request)).toMatchSnapshot();
  //jak odpalimy za pierwszym razem, to generuje snapshot, za każdym następnym sprawdza czy jest zgodny ze snapshotem, a jak się nie zgadza, to mówi coś typu "ej brakuje tutaj takiej i takiej linijki)"
  //można odpalić z flagą --u, to wtedy update'uje snapshoty
  //jest też toMatchInlineSnapshot(), ale to się czasem wywala :d wtedy wrzuca pod testem ten snapshot
});

test("fails with incorrect email", async () => {
  const request = generateRequest({ email: "mihaugmail.com" });
  expect(await validate(request)).toMatchSnapshot();
});

test("fails with empty answers", async () => {
  const request = generateRequest({ answers: null });
  expect(await validate(request)).toEqual([
    //moglibyśmy spróbować w tym arrayu cały ten obiekt, co chcemy, ale możemy sobie wybrać tylko niektóre pola, np. "{constrains: {isEmail: "email must be an email", isString: "email must be a string"}, property {email}}"
    // może być też partial matcher, np. array, w którym pierwszy ma być obiektem, który zawiera takie pola jak... wtedy będzie
    expect.objectContaining({
      constraints: {
        isEmail: "email must be an email",
        isString: "email must be a string",
      },
      property: "email",
    }),
  ]);
});

test("fails with incorrect answer", async () => {
  const request = generateRequest({ answers: [new Answer()] });
  expect(await validate(request)).toMatchSnapshot();
});
