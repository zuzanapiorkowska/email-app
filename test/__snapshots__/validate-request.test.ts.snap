// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fails with empty answers 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isString": "each value in answers must be a string",
    },
    "property": "answers",
    "target": Request {
      "answers": null,
      "email": "mihau@gmail.com",
    },
    "value": null,
  },
]
`;

exports[`fails with incorrect email 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isEmail": "email must be an email",
    },
    "property": "email",
    "target": Request {
      "email": "mihaugmail.com",
    },
    "value": "mihaugmail.com",
  },
]
`;

exports[`fails with incorrect email 2`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isEmail": "email must be an email",
    },
    "property": "email",
    "target": Request {
      "email": "mihaugmail.com",
    },
    "value": "mihaugmail.com",
  },
]
`;

exports[`fails without email 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isEmail": "email must be an email",
      "isString": "email must be a string",
    },
    "property": "email",
    "target": Request {
      "email": null,
    },
    "value": null,
  },
]
`;
